README.md
# Haskell Music Compiler
  This is a music compiler, which means the language written will be compiled into an audio file. The program handles both compiling and running of the source code
the language is unnamed at present. This language doesn't support inputs at present, which means the program should compile and run with the only result/output being
the audio file generated.

## example
```
fun main{
p (0.5 2)
p (0.5 (1 3) )
p (1.0 1)
p (0.25 3)
p (0.5 4)
p (1.0 1)
p (0.75 6)
p (0.1 7)
}
```
p stands for play, the next is a float denoting the note duration, and following is an integer if its only playing one note, or a set of integers if its playing multiple notes together.

## Building and running
### build
```cosole
$cabal build
```
### run
```console
$ cabal run filename 
$ make 
```
renaming output file isnt a feature at present. Currently `cabal run <filename>` will yeild output/outpu.bin
make command can be used to convert all bin files inside output/ to wav files ( you'll need to have ffmpeg installed )



ChangeLog.md
# Revision history for haskell-music

## 0.1.0.0  -- 2021-10-03

* First version. generates music from the given array inside the code


Main.hs
import qualified Data.ByteString.Lazy as B
import qualified Data.ByteString.Builder as B
import Data.Foldable
import System.Process
import Text.Printf

type Pulse = Float
type Seconds = Float
type Samples = Float
type Hz = Float
type Semitones = Float


outputFilePath :: FilePath
outputFilePath = "output.bin"

outputWavPath :: FilePath
outputWavPath = "output.wav"

volume::Float
volume = 0.5

sampleRate::Samples
sampleRate = 48000

pitchStandard::Hz
pitchStandard=440.0

f::Semitones->Hz
f n= pitchStandard * (2 ** (n/12.0))

freq::Hz->Seconds->[Pulse]
freq hz duration = 
    map (* volume) $ output 
    where 
        step = (hz * 2 * pi) / sampleRate
        output::[Pulse]
        output = map sin $ map (* step) [0.0 .. sampleRate * duration]

envelop::Seconds->Seconds->[Pulse]->[Pulse]
envelop attack release baseWave=
    if attack+release > 1.0 then error "not able to create attack and release"
    else
        zipWith (*) baseWave env
        where
            baseLen = fromIntegral $ length baseWave
            attackLen = attack * baseLen
            attackStep = 1.0 / attackLen
            attackArray = [0.0,attackStep..1.0]
            releaseLen = release * baseLen
            releaseStep = 1.0 / releaseLen
            releaseArray = reverse [0.0,releaseStep..1.0]
            sustainArray = take ( round (baseLen - (attackLen + releaseLen)) ) $ repeat 1.0
            env = concat [attackArray , sustainArray , releaseArray] :: [Pulse]
            

note::Semitones->Seconds->[Pulse]
note n duration = envelop 0.1 0.1 $ freq (f n) duration

wave::[Pulse]
wave = concat [note i 1.0|i<-[0..10]]


save::FilePath ->IO ()
save filePath = B.writeFile filePath $ B.toLazyByteString $ fold $ map B.floatLE wave

main::IO()
main = do
    save outputFilePath
    return ()


src/main.hs
-- for music composition
import Composer

-- For parsing input
import Compiler.Grammar
import Compiler.Tokens

-- Commnad line arguments and data input
import System.IO
import System.Environment

-- Functions regarding parsing input
readContent = do
    args <- getArgs
    if ( length args ) == 0
    then do
         contents <- getContents
         return contents
    else do
         handle <- openFile ( head args ) ReadMode
         contents <- hGetContents handle
         return contents

main::IO()
main = do
    -- reading the content from either command line or stdin
    inp <- readContent

    -- parsing the content to create ast ( form Compiler module)
    let ast = parseCalc (scanTokens inp)

    -- for testing purposes
    print ast

    -- save the binary file ( from Composer module )
    save outputFilePath ast

    --runCommand "ffmpeg -f f32le -ar 48000 -i output.bin output.wav"
    return ()


src/Composer.hs
module Composer where

import qualified Data.ByteString.Lazy as B
import qualified Data.ByteString.Builder as B
import Data.Foldable
import System.Process
import Text.Printf

import Compiler.Grammar
import Compiler.Tokens

type Pulse = Float
type Seconds = Float
type Samples = Float
type Hz = Float
type Semitones = Float


outputFilePath :: FilePath
outputFilePath = "output/output.bin"

outputWavPath :: FilePath
outputWavPath = "output/output.wav"

volume::Float
volume = 0.5

sampleRate::Samples
sampleRate = 48000

pitchStandard::Hz
pitchStandard=440.0

f::Semitones->Hz
f n= pitchStandard * (2 ** (n/12.0))

freq::Hz->Seconds->[Pulse]
freq hz duration = 
    map (* volume) $ output 
    where 
        step = (hz * 2 * pi) / sampleRate
        output::[Pulse]
        output = map sin $ map (* step) [0.0 .. sampleRate * duration]

envelop::Seconds->Seconds->[Pulse]->[Pulse]
envelop attack release baseWave=
    if attack+release > 1.0 then error "not able to create attack and release"
    else
        zipWith (*) baseWave env
        where
            baseLen = fromIntegral $ length baseWave
            attackLen = attack * baseLen
            attackStep = 1.0 / attackLen
            attackArray = [0.0,attackStep..1.0]
            releaseLen = release * baseLen
            releaseStep = 1.0 / releaseLen
            releaseArray = reverse [0.0,releaseStep..1.0]
            sustainArray = take ( round (baseLen - (attackLen + releaseLen)) ) $ repeat 1.0
            env = concat [attackArray , sustainArray , releaseArray] :: [Pulse]
            

note::Semitones->Seconds->[Pulse]
note n duration = envelop 0.1 0.1 $ freq (f n) duration

chord::[Semitones]->Seconds->[Pulse]
chord ns duration = foldr (\x y -> zipWith (+) x y) (note (head ns) duration) [ (note x duration) | x <- (tail ns) ]

wave::[Pulse]
wave = concat [note i 1.0|i<-[0..10]]

concatNotes ((NoteC duration noteval):(xs)) = (chord [(fromIntegral noteval)] (duration))++(concatNotes xs)
concatNotes ((ChordC duration nlist):(xs)) = (chord [fromIntegral x | x<-nlist] duration)++(concatNotes xs)
concatNotes [] = []

compile ast = concatNotes ast

save::FilePath->Program ->IO ()
save filePath ast = B.writeFile filePath $ B.toLazyByteString $ fold $ map B.floatLE (compile ast)


